File: openapi/openapi-generator/src/main/java/ru/tinkoff/kora/openapi/generator/javagen/ServerSecuritySchemaGenerator.java
package ru.tinkoff.kora.openapi.generator.javagen;

import com.palantir.javapoet.*;
import org.openapitools.codegen.CodegenSecurity;

import javax.lang.model.element.Modifier;
import java.util.*;

public class ServerSecuritySchemaGenerator extends AbstractJavaGenerator<Map<String, Object>> {

    @Override
    public JavaFile generate(Map<String, Object> ctx) {
        var b = TypeSpec.interfaceBuilder("ApiSecurity")
            .addModifiers(Modifier.PUBLIC)
            .addAnnotation(generated())
            .addAnnotation(Classes.module);
        var tags = new HashSet<String>();
        @SuppressWarnings("unchecked")
        var authMethods = (List<CodegenSecurity>) ctx.get("authMethods");

        tags.addAll(security.interceptorTagBySecurityRequirement.values());
        tags.addAll(security.principalExtractorTagBySecurityRequirementNames.values());

        for (var tag : tags) {
            b.addType(buildTag(tag));
        }
        for (var entry : security.principalExtractorTagBySecurityRequirementNames.entrySet()) {
            var securityRequirementNames = entry.getKey();
            var tag = entry.getValue();
            if (securityRequirementNames.size() > 1) {
                b.addType(buildAuthData(tag, securityRequirementNames, authMethods));
            }
        }

        for (var entry : security.interceptorTagBySecurityRequirement.entrySet()) {
            var requirement = entry.getKey();
            var tag = entry.getValue();
            b.addType(buildAuthGroupInterceptor(tag, new ArrayList<>(requirement), authMethods));
            b.addMethod(buildAuthGroupInterceptorComponent(tag, new ArrayList<>(requirement), authMethods));
        }

        return JavaFile.builder(apiPackage, b.build()).build();
    }

    private TypeSpec buildAuthData(String tag, Set<String> securityRequirementNames, List<CodegenSecurity> authMethods) {
        var constructor = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        for (var securityRequirementName : securityRequirementNames) {
            var authMethod = authMethods.stream().filter(m -> m.name.equals(securityRequirementName)).findFirst().get();
            var javadoc = authMethodParameterJavadoc(authMethod);
            constructor.addParameter(ParameterSpec.builder(String.class, securityRequirementName)
                .addJavadoc(javadoc)
                .build()
            );
        }
        return TypeSpec.recordBuilder(tag + "AuthData")
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC)
            .recordConstructor(constructor.build())
            .build();
    }

    private static String authMethodParameterJavadoc(CodegenSecurity authMethod) {
        if (authMethod.isOAuth || authMethod.isOpenId || authMethod.isBasic || authMethod.isBasicBasic || authMethod.isBasicBearer) {
            return "'Authorization' header of request\n";
        }
        if (authMethod.isApiKey) {
            if (authMethod.isKeyInHeader) {
                return "'" + authMethod.keyParamName + "' header of request\n";
            } else if (authMethod.isKeyInQuery) {
                return "'" + authMethod.keyParamName + "' query parameter value\n";
            } else if (authMethod.isKeyInCookie) {
                return "'" + authMethod.keyParamName + "' cookie value\n";
            } else {
                throw new IllegalArgumentException();
            }

        }
        return "\n";
    }

    private MethodSpec buildAuthGroupInterceptorComponent(String interceptorTag, List<Map<String, Set<String>>> security, List<CodegenSecurity> authMethods) {
        var interceptorClass = ClassName.get(apiPackage, "ApiSecurity", interceptorTag + "HttpServerInterceptor");
        var b = MethodSpec.methodBuilder(interceptorTag + "HttpServerInterceptor")
            .addModifiers(Modifier.PUBLIC, Modifier.DEFAULT)
            .addAnnotation(securityTagAnnotation(interceptorTag))
            .addAnnotation(Classes.defaultComponent)
            .returns(interceptorClass)
            .addCode("return new $T(", interceptorClass);
        var seen = new HashSet<String>();
        for (var securityRequirement : security) {
            var principalExtractorTag = this.security.principalExtractorTagBySecurityRequirementNames.get(securityRequirement.keySet());
            if (!seen.add(principalExtractorTag)) {
                continue;
            }
            var param = securityRequirementPrincipalExtractorParameter(authMethods, securityRequirement, principalExtractorTag);
            b.addParameter(param);
            if (seen.size() > 1) {
                b.addCode(", ");
            }
            b.addCode("$N", param.name());
        }
        b.addCode(");\n");

        return b.build();
    }

    private ParameterSpec securityRequirementPrincipalExtractorParameter(List<CodegenSecurity> authMethods, Map<String, Set<String>> securityRequirement, String principalExtractorTag) {
        var needScopes = authMethods.stream()
            .filter(auth -> securityRequirement.containsKey(auth.name))
            .anyMatch(auth -> auth.isOAuth || auth.isOpenId);
        var extractorType = ParameterizedTypeName.get(
            Classes.httpServerPrincipalExtractor,
            securityRequirement.size() == 1 ? ClassName.get(String.class) : ClassName.get(apiPackage, "ApiSecurity", principalExtractorTag + "AuthData"),
            needScopes ? Classes.principalWithScopes : Classes.principal
        );
        return ParameterSpec.builder(extractorType, principalExtractorTag)
            .addAnnotation(securityTagAnnotation(principalExtractorTag))
            .build();
    }

    private TypeSpec buildAuthGroupInterceptor(String interceptorTag, List<Map<String, Set<String>>> security, List<CodegenSecurity> authMethods) {
        var b = TypeSpec.classBuilder(interceptorTag + "HttpServerInterceptor")
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
            .addAnnotation(generated())
            .addSuperinterface(Classes.httpServerInterceptor);
        var constructor = MethodSpec.constructorBuilder()
            .addModifiers(Modifier.PUBLIC);
        var seen = new HashSet<String>();
        for (var securityRequirement : security) {
            var principalExtractorTag = this.security.principalExtractorTagBySecurityRequirementNames.get(securityRequirement.keySet());
            if (!seen.add(principalExtractorTag)) {
                continue;
            }
            var param = securityRequirementPrincipalExtractorParameter(authMethods, securityRequirement, principalExtractorTag);
            constructor.addParameter(param);
            constructor.addStatement("this.$N = $N", param.name(), param.name());
            b.addField(param.type(), param.name());
        }
        b.addMethod(constructor.build());
        var intercept = MethodSpec.methodBuilder("intercept")
            .addAnnotation(Override.class)
            .addModifiers(Modifier.PUBLIC)
            .returns(Classes.httpServerResponse)
            .addParameter(Classes.httpServerRequest, "request")
            .addParameter(Classes.httpServerInterceptChain, "chain")
            .addException(Exception.class);

        var securitySchemaSeen = new HashSet<String>();
        var securityRequirementSeen = new HashSet<String>();
        for (var securityRequirement : security) {
            for (var entry : securityRequirement.entrySet()) {
                var securitySchemaName = entry.getKey();
                var scopes = entry.getValue();
                if (securitySchemaSeen.add(securitySchemaName)) {
                    var securitySchema = authMethods.stream().filter(s -> s.name.equals(securitySchemaName)).findFirst().get();
                    if (securitySchema.isApiKey) {
                        if (securitySchema.isKeyInHeader) {
                            intercept.addStatement("var $N = request.headers().getFirst($S)", securitySchemaName, securitySchema.keyParamName);
                        } else if (securitySchema.isKeyInQuery) {
                            intercept.addStatement("var $N = request.queryParams().get($S)", securitySchemaName + "_list", securitySchema.keyParamName);
                            intercept.addStatement("var $N = $N == null || $N.isEmpty() ? null : $N.iterator().next()", securitySchemaName, securitySchemaName + "_list", securitySchemaName + "_list", securitySchemaName + "_list");
                        } else if (securitySchema.isKeyInCookie) {
                            intercept.addStatement("var $N = request.cookies().stream().filter(c -> $S.equals(c.name())).map(c -> c.value()).findFirst().orElse(null)", securitySchemaName, securitySchema.keyParamName);
                        } else {
                            throw new IllegalArgumentException();
                        }
                    } else if (securitySchema.isBasicBasic || securitySchema.isBasicBearer || securitySchema.isOAuth) {
                        intercept.addStatement("var $N = request.headers().getFirst($S)", securitySchemaName, "Authorization");
                    } else {
                        throw new IllegalArgumentException();
                    }
                }
            }
            var extractorTag = this.security.principalExtractorTagBySecurityRequirementNames.get(securityRequirement.keySet());
            if (securityRequirementSeen.add(extractorTag)) {
                if (securityRequirement.size() == 1) {
                    intercept.addStatement("var $N = this.$N.extract(request, $N)", extractorTag, extractorTag, securityRequirement.keySet().stream().findFirst().get());
                } else {
                    var authData = ClassName.get(this.apiPackage, "ApiSecurity", extractorTag + "AuthData");

                    var params = securityRequirement.keySet()
                        .stream()
                        .map(n -> CodeBlock.of("$N", n))
                        .collect(CodeBlock.joining(", ", "(", ")"));
                    intercept.addStatement("var $N = this.$N.extract(\n  request,\n  new $T$L)", extractorTag, extractorTag, authData, params);
                }
            }
            intercept.beginControlFlow("if ($N != null)", extractorTag);
            var scopesCount = 0;
            for (var entry : securityRequirement.entrySet()) {
                var securitySchemaName = entry.getKey();
                var scopes = entry.getValue();
                var securitySchema = authMethods.stream().filter(s -> s.name.equals(securitySchemaName)).findFirst().get();
                if (Boolean.TRUE.equals(securitySchema.isOAuth) && !scopes.isEmpty()) {
                    for (var scope : scopes) {
                        intercept.beginControlFlow("if ($N.scopes().contains($S))", extractorTag, scope);
                        scopesCount++;
                    }
                }
            }
            intercept.addStatement("return $T.with($N, () -> chain.process(request))", Classes.principal, extractorTag);
            for (var i = 0; i < scopesCount; i++) {
                intercept.endControlFlow();
            }

            intercept.endControlFlow();

            intercept.addCode("\n");
        }
        intercept.addStatement("throw $T.of(403, $S)", Classes.httpServerResponseException, "Forbidden");
        b.addMethod(intercept.build());
        return b.build();
    }

    private TypeSpec buildTag(String name) {
        return TypeSpec.classBuilder(name)
            .addAnnotation(generated())
            .addModifiers(Modifier.PUBLIC, Modifier.STATIC, Modifier.FINAL)
            .build();
    }
}

File: openapi/openapi-generator/src/main/kotlin/ru/tinkoff/kora/openapi/generator/kotlingen/ServerSecuritySchemaGenerator.kt
package ru.tinkoff.kora.openapi.generator.kotlingen

import com.squareup.kotlinpoet.*
import com.squareup.kotlinpoet.ParameterizedTypeName.Companion.parameterizedBy
import org.openapitools.codegen.CodegenSecurity


class ServerSecuritySchemaGenerator : AbstractKotlinGenerator<Map<String, Any>>() {
    override fun generate(ctx: Map<String, Any>): FileSpec {
        val b = TypeSpec.interfaceBuilder("ApiSecurity")
            .addAnnotation(generated())
            .addAnnotation(Classes.module.asKt())
        val tags = mutableSetOf<String>()
        val authMethods = ctx["authMethods"] as List<CodegenSecurity>

        tags.addAll(security.interceptorTagBySecurityRequirement.values)
        tags.addAll(security.principalExtractorTagBySecurityRequirementNames.values)

        for (tag in tags) {
            b.addType(buildTag(tag))
        }
        for (entry in security.principalExtractorTagBySecurityRequirementNames.entries) {
            val securityRequirementNames = entry.key
            val tag = entry.value
            if (securityRequirementNames.size > 1) {
                b.addType(buildAuthData(tag, securityRequirementNames, authMethods))
            }
        }
        for (entry in security.interceptorTagBySecurityRequirement.entries) {
            val requirement = entry.key
            val tag = entry.value
            b.addType(buildAuthGroupInterceptor(tag, requirement.toList(), authMethods))
            b.addFunction(buildAuthGroupInterceptorComponent(tag, requirement.toList(), authMethods))
        }

        return FileSpec.get(apiPackage, b.build())
    }

    private fun buildAuthData(tag: String, securityRequirementNames: Set<String>, authMethods: List<CodegenSecurity>): TypeSpec {
        val constructor = FunSpec.constructorBuilder()
        val type = TypeSpec.classBuilder(tag + "AuthData")
            .addModifiers(KModifier.DATA)
        for (securityRequirementName in securityRequirementNames) {
            val authMethod = authMethods.first { m -> m.name == securityRequirementName }
            val doc = authMethodParameterJavadoc(authMethod)
            type.addProperty(
                PropertySpec.builder(securityRequirementName, String::class.asClassName().copy(nullable = true))
                    .addKdoc(doc)
                    .initializer("%N", securityRequirementName)
                    .build()
            )
            constructor.addParameter(
                ParameterSpec.builder(securityRequirementName, String::class.asClassName().copy(nullable = true))
                    .addKdoc(doc)
                    .build()
            )
        }
        return type.primaryConstructor(constructor.build()).build()
    }


    private fun authMethodParameterJavadoc(authMethod: CodegenSecurity): String {
        if (authMethod.isOAuth || authMethod.isOpenId || authMethod.isBasic || authMethod.isBasicBasic || authMethod.isBasicBearer) {
            return "'Authorization' header of request\n"
        }
        if (authMethod.isApiKey) {
            return when {
                authMethod.isKeyInHeader -> "'" + authMethod.keyParamName + "' header of request\n"
                authMethod.isKeyInQuery -> "'" + authMethod.keyParamName + "' query parameter value\n"
                authMethod.isKeyInCookie -> "'" + authMethod.keyParamName + "' cookie value\n"
                else -> throw IllegalArgumentException()
            }
        }
        return "\n"
    }


    private fun buildAuthGroupInterceptor(interceptorTag: String, security: List<Map<String, Set<String>>>, authMethods: List<CodegenSecurity>): TypeSpec {
        val b = TypeSpec.classBuilder(interceptorTag + "HttpServerInterceptor")
            .addAnnotation(generated())
            .addSuperinterface(Classes.httpServerInterceptor.asKt())
        val constructor = FunSpec.constructorBuilder()
        val seen = hashSetOf<String>()
        for (securityRequirement in security) {
            val principalExtractorTag = this.security.principalExtractorTagBySecurityRequirementNames[securityRequirement.keys]!!
            if (!seen.add(principalExtractorTag)) {
                continue
            }
            val param = securityRequirementPrincipalExtractorParameter(authMethods, securityRequirement, principalExtractorTag)
            constructor.addParameter(param)
            b.addProperty(
                PropertySpec.builder(param.name, param.type, KModifier.PRIVATE)
                    .initializer("%N", param.name)
                    .build()
            )
        }
        b.primaryConstructor(constructor.build())
        val intercept = FunSpec.builder("intercept")
            .addModifiers(KModifier.OVERRIDE)
            .returns(Classes.httpServerResponse.asKt())
            .addParameter("request", Classes.httpServerRequest.asKt())
            .addParameter("chain", Classes.httpServerInterceptChain.asKt())

        val securitySchemaSeen = mutableSetOf<String>()
        val securityRequirementSeen = mutableSetOf<String>()
        for (securityRequirement in security) {
            for ((securitySchemaName, scopes) in securityRequirement.entries) {
                if (securitySchemaSeen.add(securitySchemaName)) {
                    val securitySchema = authMethods.first { s -> s.name.equals(securitySchemaName) }
                    if (securitySchema.isApiKey) {
                        when {
                            securitySchema.isKeyInHeader -> intercept.addStatement(
                                "val %N = request.headers().getFirst(%S)",
                                securitySchemaName,
                                securitySchema.keyParamName
                            )

                            securitySchema.isKeyInQuery -> intercept.addStatement(
                                "val %N = request.queryParams().get(%S)?.firstOrNull()",
                                securitySchemaName,
                                securitySchema.keyParamName
                            )

                            securitySchema.isKeyInCookie -> intercept.addStatement(
                                "val %N = request.cookies().firstOrNull { c -> %S == c.name() }?.value()",
                                securitySchemaName,
                                securitySchema.keyParamName
                            )

                            else -> throw IllegalArgumentException()
                        }
                    } else if (securitySchema.isBasicBasic || securitySchema.isBasicBearer || securitySchema.isOAuth) {
                        intercept.addStatement("var %N = request.headers().getFirst(%S)", securitySchemaName, "Authorization")
                    } else {
                        throw IllegalArgumentException()
                    }
                }
            }
            val extractorTag = this.security.principalExtractorTagBySecurityRequirementNames[securityRequirement.keys]!!
            if (securityRequirementSeen.add(extractorTag)) {
                if (securityRequirement.size == 1) {
                    intercept.addStatement("val %N = this.%N_.extract(request, %N)", extractorTag, extractorTag, securityRequirement.keys.first())
                } else {
                    val authData = ClassName(this.apiPackage, "ApiSecurity", extractorTag + "AuthData");
                    val params = securityRequirement.keys.map { CodeBlock.of("%N", it) }.joinToCode(", ", "(", ")")
                    intercept.addStatement("val %N = this.%N_.extract(\n  request,\n  %T%L)", extractorTag, extractorTag, authData, params)
                }
            }
            intercept.beginControlFlow("if (%N != null)", extractorTag)
            var scopesCount = 0

            for ((securitySchemaName, scopes) in securityRequirement.entries) {
                val securitySchema = authMethods.first { it.name.equals(securitySchemaName) }
                if (securitySchema.isOAuth == true && !scopes.isEmpty()) {
                    for (scope in scopes) {
                        intercept.beginControlFlow("if (%N.scopes().contains(%S))", extractorTag, scope)
                        scopesCount++
                    }
                }
            }
            intercept.addStatement("return %T.with<%T, %T>(%N) { chain.process(request) }", Classes.principal.asKt(), Classes.httpServerResponse.asKt(), RuntimeException::class.asClassName(), extractorTag)
            for (i in 0 until scopesCount) {
                intercept.endControlFlow()
            }

            intercept.endControlFlow()

            intercept.addCode("\n");
        }

        intercept.addStatement("throw %T.of(403, %S)", Classes.httpServerResponseException.asKt(), "Forbidden")
        b.addFunction(intercept.build())
        return b.build()
    }


    private fun buildAuthGroupInterceptorComponent(interceptorTag: String, security: List<Map<String, Set<String>>>, authMethods: List<CodegenSecurity>): FunSpec {
        val interceptorClass = ClassName(apiPackage, "ApiSecurity", interceptorTag + "HttpServerInterceptor")
        val b = FunSpec.builder(interceptorTag + "HttpServerInterceptorComponent")
            .addAnnotation(securityTagAnnotation(interceptorTag))
            .addAnnotation(Classes.defaultComponent.asKt())
            .returns(interceptorClass)
            .addCode("return %T(", interceptorClass)
        val seen = mutableSetOf<String>()
        for (securityRequirement in security) {
            val principalExtractorTag = this.security.principalExtractorTagBySecurityRequirementNames.get(securityRequirement.keys)!!
            if (!seen.add(principalExtractorTag)) {
                continue
            }
            val param = securityRequirementPrincipalExtractorParameter(authMethods, securityRequirement, principalExtractorTag)
            b.addParameter(param)
            if (seen.size > 1) {
                b.addCode(", ")
            }
            b.addCode("%N", param.name)
        }
        b.addCode(")\n")

        return b.build()
    }

    private fun securityRequirementPrincipalExtractorParameter(
        authMethods: List<CodegenSecurity>,
        securityRequirement: Map<String, Set<String>>,
        principalExtractorTag: String
    ): ParameterSpec {
        val needScopes = authMethods
            .filter { auth -> securityRequirement.containsKey(auth.name) }
            .any { auth -> auth.isOAuth || auth.isOpenId }
        val extractorParam = if (securityRequirement.size == 1) String::class.asClassName() else ClassName(apiPackage, "ApiSecurity", principalExtractorTag + "AuthData")
        val principalType = if (needScopes) Classes.principalWithScopes.asKt() else Classes.principal.asKt()
        val extractorType = Classes.httpServerPrincipalExtractor.asKt().parameterizedBy(extractorParam, principalType)
        return ParameterSpec.builder(principalExtractorTag + "_", extractorType)
            .addAnnotation(securityTagAnnotation(principalExtractorTag))
            .build()
    }



    private fun buildTag(name: String): TypeSpec {
        return TypeSpec.classBuilder(name)
            .addAnnotation(generated())
            .build()
    }
}
